File: apps/java/app-50-energy-billing/src/main/java/com/energy/billing/support/ReferenceGuards.java
package com.energy.billing.support;

import java.net.URI;
import java.util.Locale;
import java.util.Set;

public final class ReferenceGuards {
    private ReferenceGuards() {
    }

    public static boolean sameOwner(Object recordOwner, Object currentUser) {
        return String.valueOf(recordOwner).equals(String.valueOf(currentUser));
    }

    public static boolean allowedCallback(String target, Set<String> allowedHosts) {
        try {
            URI parsed = URI.create(target == null ? "" : target);
            String scheme = parsed.getScheme();
            return ("http".equals(scheme) || "https".equals(scheme)) && allowedHosts.contains(parsed.getHost());
        } catch (IllegalArgumentException ex) {
            return false;
        }
    }

    public static String normalizeIdentifier(Object value) {
        return String.valueOf(value == null ? "" : value).trim().toLowerCase(Locale.ROOT);
    }
}

File: apps/javascript/app-17-iot-dashboard/src/referenceGuards.js
function sameOwner(recordOwner, currentUser) {
  return String(recordOwner) === String(currentUser);
}

function allowedCallback(target, allowedHosts) {
  try {
    const parsed = new URL(target);
    return ["http:", "https:"].includes(parsed.protocol) && allowedHosts.includes(parsed.hostname);
  } catch (_error) {
    return false;
  }
}

function normalizeIdentifier(value) {
  return String(value ?? "").trim().toLowerCase();
}

module.exports = { sameOwner, allowedCallback, normalizeIdentifier };
